asynq/tools.py

```diff
@@ -305,3 +307,39 @@ def resume(self):
 
     def pause(self):
         self.total_time += utime() - self._last_start_time
+
+
+class AsyncEventHook(EventHook):
+    """EventHook that supports async handlers.
+
+    When the event triggers, all the handlers will be invoked asynchronously using async_call.
+
+    Since it uses async_call, it supports normal handlers by default
+    (they don't need to be @async() decorated).
+
+    """
+
+    @async()
+    def trigger(self, *args):
+        yield [async_call.async(handler, *args) for handler in self]
+
+    @async()
+    def safe_trigger(self, *args):
+        wrapped_handlers = [self._create_safe_wrapper(handler) for handler in self]
+        results = yield [wrapped_handler.async(*args) for wrapped_handler in wrapped_handlers]
+        results = filter(None, results)
+        if len(results) > 0:
```

Comment:
This doesn't work in Python 3, because filter() returns a lazy iterator instead of a sequence. A possible alternative implementation is

```
for error in filter(None, results):
    reraise(error)
```